code/game/machinery/computer/HolodeckControl.dm

```diff
@@ -22,22 +24,34 @@
 /obj/machinery/computer/HolodeckControl/attack_paw(var/mob/user as mob)
 	return
 
-///obj/machinery/computer/HolodeckControl/attack_ghost(mob/dead/observer/user)
-//	if(!ticker || ticker.current_state != GAME_STATE_PLAYING)
-//		to_chat(user, "<span class='notice'>You can't do this until the game has started.</span>")
-//		return
-//	if(!linkedholodeck)
-//		return
-//	var/turf/spawnturf
-//	var/list/L = get_area_turfs(linkedholodeck.type)
-//	L = shuffle(L)
-//	for(var/turf/T in L)
-//		if(!is_blocked_turf(T))
-//			spawnturf = T
-//			break
-//	if(spawnturf)
-//		user.forceMove(spawnturf)
-//		user.transmogrify(/mob/living/simple_animal/hologram/advanced, TRUE)
+/obj/machinery/computer/HolodeckControl/proc/spawn_holoperson(mob/dead/observer/user)
+	if(stat & (NOPOWER|BROKEN|MAINT))
+		return
+	if(!ticker || ticker.current_state != GAME_STATE_PLAYING)
+		to_chat(user, "<span class='notice'>You can't do this until the game has started.</span>")
+		return
+	if(!linkedholodeck)
+		return
+	if(connected_holopeople.len >=4)
+		to_chat(user, "<span class='notice'>\The [src] cannot sustain any additional advanced holograms. Please try again when there are fewer advanced holograms on the holodeck.</span>")
+		return
+	var/turf/spawnturf
+	var/list/L = get_area_turfs(linkedholodeck.type)
+	L = shuffle(L)
```

Comment:
Using a while loop with a pick is probably a better solution.

Reply:
Shuffling and then cycling through ensures both that a random turf is chosen and that no turf is checked twice. Using `pick()` in a loop could result in unnecessary duplicate checks.

Reply:
The shuffle itself would FAR outweigh the performance lost from a few rare duplicates.